core/node/rpc/syncv3/eventbus/eventbus.go

```diff
@@ -0,0 +1,498 @@
+	// from future updates on the stream. It is expected that this update is sent to the client
+	// and that the client will resubscribe with a backoff mechanism.
+	OnUpdate(update *SyncStreamsResponse)
+}
+
+// StreamSubscriptionManager allows subscribers to subscribe and unsubscribe on/from stream updates.
+type StreamSubscriptionManager interface {
+	// EnqueueSubscribe on stream events.
+	//
+	// Implementation must make this a non-blocking operation.
+	//
+	// This can fail with (Err_NOT_FOUND) when the stream is not found.
+	// If the subscriber is already subscribed to the stream, this is a noop.
+	//
+	// When the subscriber receives a SyncOp_SYNC_DOWN stream update the subscriber is automatically
+	// unsubscribed from the stream and won't receive further updates.
+	EnqueueSubscribe(cookie *SyncCookie, subscriber StreamSubscriber) error
+
+	// EnqueueUnsubscribe from events.
+	//
+	// Implementation must make this a non-blocking operation.
+	//
+	// Note that unsubscribing happens automatically when the subscriber
+	// receives a SyncOp_SYNC_DOWN update for the stream.
+	//
+	// If the subscriber is not subscribed to the stream, this is a noop.
+	EnqueueUnsubscribe(streamID StreamId, subscriber StreamSubscriber) error
+
+	// EnqueueBackfill requests a backfill message.
+	//
+	// The target syncer will request a stream update message by the given cookie and send it
+	// back to the target sync operation through the given chain of sync IDs.
+	EnqueueBackfill(cookie *SyncCookie, syncIDs ...string) error
+}
+
+type eventBusMessageStreamUpdate struct {
+	msg     *SyncStreamsResponse
+	version int
+}
+
+type eventBusMessageSub struct {
+	cookie     *SyncCookie
+	subscriber StreamSubscriber
+}
+
+type eventBusMessageUnsub struct {
+	streamID   StreamId
+	subscriber StreamSubscriber
+}
+
+type eventBusMessageBackfill struct {
+	cookie  *SyncCookie
+	syncIDs []string
+}
+
+// eventBusMessage is put on the internal event bus queue for processing in eventBusImpl.run.
+type eventBusMessage struct {
+	update   *eventBusMessageStreamUpdate
+	sub      *eventBusMessageSub
+	unsub    *eventBusMessageUnsub
+	backfill *eventBusMessageBackfill
+}
+
+// eventBusImpl is a concrete implementation of the StreamUpdateEmitter and StreamSubscriber interfaces.
+// It is responsible for handling stream updates from syncer.StreamUpdateEmitter and distributes updates
+// to subscribers. As such it keeps track of which subscribers are subscribed on updates on which streams.
+type eventBusImpl struct {
+	log *logging.Log
+	// TODO: discuss if we want to have an unbounded queue here?
+	// Processing items on the queue should be fast enough to always keep up with the added items.
+	queue    *dynmsgbuf.DynamicBuffer[*eventBusMessage]
+	registry syncer.Registry
+	// subscribers is the list of subscribers grouped by stream ID and syncer version.
+	subscribers map[StreamId]map[int][]StreamSubscriber
```

Comment:
This map must be protected against concurrent access or replaced with a thread-safe alternative. It is modified in `OnStreamEvent` and in a `processXX` methods called in the background task that executes the `run` method.